src/super_gradients/training/utils/quantization_utils.py

```diff
@@ -0,0 +1,329 @@
+            profile0 = builder.create_optimization_profile()
+            for name, shape in zip(input_names, input_shapes):
+                profile0.set_shape(
+                    name, min=(1, *shape[1:]), opt=(trt_max_batch_size, *shape[1:]),
+                    max=(trt_max_batch_size, *shape[1:])
+                )
+            config.add_optimization_profile(profile0)
+
+            # Additional builder_config flags can be set prior to building the engine
+            if quantization_level == QuantizationLevel.FP16:
+                config.set_flag(trt.BuilderFlag.FP16)
+                try:
+                    builder.fp16_mode = True
+                except:
+                    # TRT8 breaking API - supporting older versions
+                    pass
+            elif quantization_level == QuantizationLevel.INT8:
+                config.set_flag(trt.BuilderFlag.INT8)
+                config.set_flag(trt.BuilderFlag.STRICT_TYPES)
+
+            # Compilation parameters
+            config.max_workspace_size = 4 << 30
+            try:
+                builder.max_workspace_size = 4 << 30
+            except:
+                # TRT8 breaking API - supporting older versions
+                pass
+
+            # Set max batch size
+            builder.max_batch_size = trt_max_batch_size
+
+            engine = builder.build_engine(network, config)
+
+            return engine
+
+
+def save_trt_engine_from_onnx_ckpt(onnx_ckpt_path, output_engine_path, quantization_level=QuantizationLevel.INT8,
+                                   max_batch_size=1):
+    """
+    Method for creating trt eingine from an exisiting ONNX file
+
+    :param onnx_ckpt_path: str, path to ONNX file
+    :param output_engine_path: str, target path for the trt engine file.
+    :param quantization_level: QuantizationLevel, quantization level for the engine (should be INT8 when QAT)
+    :param max_batch_size: int, trt maximal batch size
+    """
+    if os.path.exists(onnx_ckpt_path):
+        print("Building engine from file {}".format(onnx_ckpt_path))
+        trt_engine = build_trt_engine_from_onnx_ckpt(onnx_ckpt_path=onnx_ckpt_path,
+                                                     trt_max_batch_size=max_batch_size,
+                                                     quantization_level=quantization_level)
+        serialized_engine = bytes(trt_engine.serialize())
+        engine_dict = {'engine': serialized_engine,
+                       'compiler': 'trt',
+                       'precision': quantization_level.value}
+        with open(output_engine_path, "wb") as f:
+            pickle.dump(engine_dict, f)
+            print(
+                f'Successfully converted {onnx_ckpt_path} to {output_engine_path} with max batch size {max_batch_size} and {quantization_level} quantization.')
+    else:
+        raise ValueError("The input file does not exist.")
+
+
+class Int8CalibrationPreTrainingCallback(PhaseCallback):
```

Comment:
why do we need this? do we have to calibrate before QAT?

Reply:
We go to have the quantised module statistics collected, yes. Otherwise the QAT won't make much sense....
So either we do it here, or initialise the quant modules according to the specified statistics type and load a calibrated checkpoint (i.e one that holds the statistics/ amax already, with quant modules).